daemon/policy.go

```diff
@@ -301,16 +390,123 @@ func (d *Daemon) PolicyAdd(rules policyAPI.Rules, opts *AddOptions) (newRev uint
 		d.SendNotification(monitorAPI.AgentNotifyPolicyUpdated, repr)
 	}
 
-	return newRev, nil
+	// Only regenerate endpoints which are needed to be regenerated as a result
+	// of the rule update. The rules which were imported most likely do not select
+	// all endpoints in the policy repository (and may not select any at all).
+	// The "reacting" to rule updates enqueues events for all endpoints. Once
+	// all endpoints have events queued up, this function will return.
+
+	r := &PolicyReactionEvent{
+		d:                d,
+		wg:               &policySelectionWG,
+		eps:              policyEps,
+		endpointsToRegen: endpointsToRegen,
+		newRev:           newRev,
+	}
+
+	ev := eventqueue.NewEvent(r)
+	// This event may block if the RuleReactionQueue is full. We don't care
+	// about when it finishes, just that the work it does is done in a serial
+	// order.
+	d.policy.RuleReactionQueue.Enqueue(ev)
+
+	return
+}
+
+// PolicyReactionEvent is an event which needs to be serialized after changes
+// to a policy repository for a daemon. This currently consists of endpoint
+// regenerations / policy revision incrementing for a given endpoint.
+type PolicyReactionEvent struct {
+	d                *Daemon
+	wg               *sync.WaitGroup
+	eps              []policy.Endpoint
+	endpointsToRegen *policy.IDSet
+	newRev           uint64
+}
+
+// Handle implements pkg/eventqueue/EventHandler interface.
+func (r *PolicyReactionEvent) Handle(res chan interface{}) {
+	r.d.ReactToRuleUpdates(r.wg, r.eps, r.endpointsToRegen, r.newRev)
 }
 
-// PolicyDelete deletes the policy set in the given path from the policy tree.
-// If cover256Sum is set it finds the rule with the respective coverage that
-// rule from the node. If the path's node becomes ruleless it is removed from
-// the tree.
+// ReactToRuleUpdates waits until wg is complete to do the following
+// * regenerate all endpoints in epsToRegen
+// * bump the policy revision of all endpoints not in epsToRegen, but which are
+//   in allEps, to revision rev.
+func (d *Daemon) ReactToRuleUpdates(wg *sync.WaitGroup, allEps []policy.Endpoint, epsToRegen *policy.IDSet, rev uint64) {
+	// Wait until we have calculated which endpoints need to be selected
+	// across multiple goroutines.
+	wg.Wait()
+
+	epsToRegen.Mutex.RLock()
+	defer epsToRegen.Mutex.RUnlock()
+
+	var enqueueWaitGroup sync.WaitGroup
+
+	// If an endpoint is not in the list of endpoints to regenerate, then the
+	// policy revision for the endpoint needs to be bumped so we can reflect that
+	// said endpoint realizes the state of the policy repository at revision rev.
+	for _, ep := range allEps {
+		if ep == nil {
+			continue
+		}
+		epID := ep.GetID16()
+		if _, ok := epsToRegen.IDs[epID]; !ok {
+			enqueueWaitGroup.Add(1)
+			go ep.PolicyRevisionBumpEvent(rev, &enqueueWaitGroup)
```

Comment:
nit: Passing the waitgroup to another function shares the responsibility which makes it more difficult to validate. If the rest of the WaitGroup is in this function, it seems much easier to understand if its usage is constrained to the one function. In this case, it's fine because the callee here is super simple, but something to keep in mind for reviewers in future.

Reply:
An upcoming commit will remove the `WaitGroup` from `PolicyRevisionBumpEvent`.